### DISCUSSION TITLE
Intermittent "entity does not exist" panic

### DISCUSSION THREAD

**Discussion by sephelps** [bevyengine/bevy#4454] [Q&A] [answered]
My implementation of a loading screen is experiencing an intermittent panic:

> thread 'Compute Task Pool (14)' panicked at 'Attempting to create an EntityCommands for entity 2v0, which doesn't exist.', /home/sephelps/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_transform-0.6.0/src/hierarchy/hierarchy_maintenance_system.rs:69:18

1. Conceptually, when entering the `Loading` state, I spawn all the entities for the loading screen--like the progress bar.
2. On update during the `Loading` state, I check to see if the assets are loaded, and if so, set the state to the next state.
3. On exit of the `Loading` state, I despawn the entities created when entering the `Loading` state.

Not every time, but sometimes, when attempting to despawn the loading screen entities, it claims that they don't exist, even though I'm getting the entity ID from a `Query` in the same function.

When the panic occurs, I also get this warning:
> 2022-04-11T03:27:29.890556Z  WARN bevy_ui::flex: Styled child in a non-UI entity hierarchy. You are using an entity with UI components as a child of an entity without UI components, results may be unexpected.

I'm assuming it has to do with parallel system running, but I thought the `on_enter`, `on_exit`, etc. were hard boundaries. Even when specifying labels and `after`, I still receive the panic sometimes. Loading a sufficiently large amount of assets stops the error.

Here is the structure of the app's systems:
```
impl Plugin for LoadingPlugin {
    fn build(&self, app: &mut App) {
        let spawn = SystemSet::on_enter(AppState::Loading)
            .with_system(spawn_loading);

        let update = SystemSet::on_update(AppState::Loading)
            .with_system(update_loading_progress)
            .with_system(check_loading_done);

        let unload = SystemSet::on_exit(AppState::Loading)
            .with_system(unload_loading);

        app.add_system_set(spawn)
            .add_system_set(update)
            .add_system_set(unload);
    }
}
```
Here is the entity spawning system:
```
fn spawn_loading(
    mut commands: Commands,
) {
    let root = commands.spawn_bundle(NodeBundle {
        style: Style {
            size: Size::new(Val::Percent(100.), Val::Percent(100.)),
            align_items: AlignItems::Center,
            justify_content: JustifyContent::Center,
            ..Default::default()
        },
        color: Color::NONE.into(),
        ..Default::default()
    }).insert(UiRoot).id();

    let container = commands.spawn_bundle(NodeBundle {
        style: Style {
            size: Size::new(Val::Px(200.), Val::Px(15.)),
            ..Default::default()
        },
        color: Color::BLACK.into(),
        ..Default::default()
    }).insert(Parent(root)).id();

    commands.spawn_bundle(NodeBundle {
        style: Style {
            size: Size::new(Val::Percent(0.), Val::Percent(100.)),
            ..Default::default()
        },
        ..Default::default()
    }).insert(UiProgress).insert(Parent(container));
}
```
Here is the function that triggers the `on_exit` `SystemSet`:
```
fn check_loading_done(
    loading: Res<LoadingAssets>,
    server: Res<AssetServer>,
    mut state: ResMut<State<AppState>>,
    next_state: Res<NextState>,
) {
    if server.get_group_load_state(loading.0.iter().map(|h| h.id)) == LoadState::Loaded {
        state.set(next_state.0);
    }
}
```
The system where this intermittent panic occurs:
```
fn unload_loading(
    mut commands: Commands,
    mut ui_query: Query<Entity, With<UiRoot>>,
) {
    // how is it possible `ui_query.single()` returns an entity but `commands.entity()` thinks it doesn't exist (and only sometimes)?
    commands.entity(ui_query.single()).despawn_recursive();
    // ...
}
```

**Comment by DJMcNab** [comment]
This seems likely to be a duplicate of https://github.com/bevyengine/bevy/issues/3195

**Reply by sephelps** [reply]
Unfortunately it happens when I despawn a single entity as well, but I have figured out a fix. Thanks!

**Comment by sephelps** [comment, marked as answer]
I was able to fix this by making the spawning and despawning systems exclusive:

```
impl Plugin for LoadingPlugin {
    fn build(&self, app: &mut App) {
        let spawn = SystemSet::on_enter(AppState::Loading)
            .with_system(spawn_loading.exclusive_system());

        let update = SystemSet::on_update(AppState::Loading)
            .with_system(update_loading_progress)
            .with_system(check_loading_done);

        let unload = SystemSet::on_exit(AppState::Loading)
            .with_system(unload_loading.exclusive_system());

        app.add_system_set(spawn)
            .add_system_set(update)
            .add_system_set(unload);
    }
}
```
